IdentifiabilityODE.mpl

```diff
@@ -557,3 +587,168 @@ CompareDiffVar := proc(dvl, dvr, var_list)
+  # this is a flag array: if i-th position == 1 then we must differentiat i-th y(t) function 
+  differentiate_ := [seq(1, i=1..nops(y_eqs))]: 
+
+  for j from 1 to s + 1 do
+    # begin differentiation
+    current_level := current_level + 1:
+    continue:=true:
+
+    for i from 1 to m do
+      if differentiate_[i]=1 then 
+        poly_d := numer(lhs(y_eqs[i]) - rhs(y_eqs[i])):
+        poly_d := Differentiate(poly_d, all_vars):
+        leader := MakeDerivative(y_vars[i], j):
+        separant := diff(poly_d, leader):
+        poly_d := simplify(leader - subs(x_eqs, -(poly_d - separant * leader) / separant)):
+        y_eqs[i]:= leader = simplify(poly_d - leader);
+        candidates := select(x-> not (GetOrderVar(x)[1]  in y_vars), indets(y_eqs[i])):
+        if op(map(x->not assigned(visibility_table[get_state_name(x, x_vars, mu)]), candidates)) <> NULL then
+          continue := foldl(`or`, op(map(x->not assigned(visibility_table[get_state_name(x, x_vars, mu)]), candidates))):
+        else
+          continue := false;
+        fi:
+        if continue then
+          differentiate_[i]:=1:
+        else
+          differentiate_[i]:=0:
+        fi:
+        for each in candidates do
+          if not assigned(visibility_table[get_state_name(each, x_vars, mu)]) then 
+              visibility_table[get_state_name(each, x_vars, mu)] := current_level:
+          fi:
+        od;
+      fi:
+    od:
+    if add(k, k in differentiate_)=0 then
+        break:
+    fi:
+  od:
+  return visibility_table;
+end proc:
+
+SubsByDepth := proc(sigma, poly_system, poly_vars, non_id, {trdegsub:=false})
+  local counting_table_states, min_count, vts, rhs_terms, max_possible,
+        rhs_term, indets_, term, substitutions, each, alg_indep, original_et_hat,
+        all_subs, names, selection, other, all_odes, each_ode;
+
+  vts := GetMinLevelBFS(sigma):
+  substitutions := table([]);
+  all_odes := map(x->expand(rhs(x)), select(f->is_diff(lhs(f)), sigma));
+  rhs_terms := []:
+  for each_ode in all_odes do
+    if whattype(each_ode) in [`+`,`*`,`^`] then
+      rhs_terms := [op(rhs_terms), op(each_ode)];
+    end if;
+    if whattype(each_ode) in [function] then
+     rhs_terms := [op(rhs_terms), (each_ode)]; 
+    end if;
+  end do;
+  
+  max_possible := max(map(rhs, [entries(vts, `pairs`)]));
+  for rhs_term in rhs_terms do
+    indets_ := convert(indets(rhs_term) minus {t}, list):
+    for term in indets_ do
+      if is_function(term) then
+        if assigned(vts[FunctionToVariable(term)]) then
+          substitutions[FunctionToVariable(term)] := vts[FunctionToVariable(term)]+1:
+        end if;
+      else
+        if not term in non_id and vts[term]=max_possible and assigned(vts[term]) then # 
+          substitutions[term] := vts[term]+1:
+        end if;
+      end if:
+    end do;
+  end do:
+  substitutions[z_aux]:=min(3, max_possible):

```

Comment:
I am not sure I understand what happens in the function up to this line. Is this all really necessary?

Reply:
After the BFS process is complete, the function carefully collects all operands in the rhs of ODEs. Then, for each variable in each rhs term, we assign weight based on visibility table (vts).